app/questionnaire/navigation.py

```diff
@@ -0,0 +1,170 @@
+import copy
+import logging
+
+from collections import defaultdict
+
+from app.helpers.schema_helper import SchemaHelper
+from app.questionnaire.location import Location
+from app.questionnaire.rules import evaluate_repeat
+
+logger = logging.getLogger(__name__)
+
+
+class Navigation(object):
+    """
+    Reads navigation config from the schema and returns collections of dicts that describe
+    completion status of each group
+    """
+    def __init__(self, survey_json, answer_store, metadata=None, completed_blocks=None):
+        self.survey_json = survey_json
+        self.metadata = metadata or {}
+        self.answer_store = answer_store
+        self.completed_blocks = completed_blocks or []
+
+    def build_navigation(self, current_group_id, current_group_instance):
+        """
+        Build navigation based on the current group/instance selected
+
+        :param current_group_id:
+        :param current_group_instance:
+        :return:
+        """
+        if 'navigation' not in self.survey_json:
+            return None
+
+        navigation = []
+
+        for group in filter(lambda x: 'hide_in_navigation' not in x, self.survey_json['groups']):
+            first_location = Location(group['id'], 0, group['blocks'][0]['id'])
+
+            logger.debug("Building frontend navigation for group %s", group)
+
+            repeating_rule = SchemaHelper.get_repeat_rule(group)
+
+            if repeating_rule:
+                navigation += self._build_repeating_navigation(repeating_rule, group, current_group_id,
+                                                               current_group_instance)
+            else:
+                navigation.append(self._build_single_navigation(group, current_group_id, first_location))
```

Comment:
This can't be consistent as one returns a list the other returns a single item as a dict. += tries to merge the dict as a result and wipes it.

It' is possible to cast the single item to a list and then add it as a list using +=, but that approach involves an unnecessary cast.